public/app/features/alerting/unified/rule-list/hooks/prometheusGroupsGenerator.ts

```diff
@@ -15,27 +14,68 @@ interface FetchGroupsOptions {
 }
 
 export function usePrometheusGroupsGenerator() {
+  const dispatch = useDispatch();
   const [getGroups] = useLazyGetGroupsQuery();
 
   return useCallback(
     async function* (ruleSource: DataSourceRulesSourceIdentifier, groupLimit: number) {
-      const getRuleSourceGroups = (options: FetchGroupsOptions) =>
-        getGroups({ ruleSource: { uid: ruleSource.uid }, ...options });
-
-      yield* genericGroupsGenerator(getRuleSourceGroups, groupLimit);
+      const getRuleSourceGroupsWithCache = async (options: FetchGroupsOptions) => {
+        const response = await getGroups({ ruleSource: { uid: ruleSource.uid }, ...options }).unwrap();
+
+        response.data.groups.forEach((group) => {
+          dispatch(
+            prometheusApi.util.upsertQueryData(
+              'getGroups',
+              { ruleSource: { uid: ruleSource.uid }, namespace: group.file, groupName: group.name },
+              { data: { groups: [group] }, status: 'success' }
+            )
+          );
+        });
+
+        return response;
+      };
+
+      yield* genericGroupsGenerator(getRuleSourceGroupsWithCache, groupLimit);
     },
-    [getGroups]
+    [getGroups, dispatch]
   );
 }
 
 export function useGrafanaGroupsGenerator() {
+  const dispatch = useDispatch();
   const [getGrafanaGroups] = useLazyGetGrafanaGroupsQuery();
 
+  const getGroupsAndProvideCache = useCallback(
+    async (options: FetchGroupsOptions) => {
+      const response = await getGrafanaGroups(options).unwrap();
+
+      // This is not mandatory to preload ruler rules, but it improves the UX
+      // Because the user waits a bit longer for the initial load but doesn't need to wait for each group to be loaded
+      const cacheAndRulerPreload = response.data.groups.map(async (group) => {
+        await dispatch(
+          alertRuleApi.endpoints.getGrafanaRulerGroup.initiate({ folderUid: group.folderUid, groupName: group.name })
+        );
+        await dispatch(
+          prometheusApi.util.upsertQueryData(
+            'getGrafanaGroups',
+            { folderUid: group.folderUid, groupName: group.name },
+            { data: { groups: [group] }, status: 'success' }
+          )
+        );
+      });
+
+      await Promise.allSettled(cacheAndRulerPreload);
+
```

Comment:
This is to be discussed. I've added ruler preloading here. It means we need to wait a bit longer for a group to render, but we don't need to wait for loading ruler when the user expands the group section - everything is already in RTKQ cache, so no loading state in the UI

Reply:
given the groups are paginated, this might not impact too much? I mean having to wait for these groups been rendered.

Reply:
> will we have loaded the ruler definitions for the rule groups they aren't interested in?
Yes